src/piadm.sh

```diff
@@ -578,44 +572,43 @@ activate() {
 		fi
 	fi
 
-	vecho "    boot image " $(cat etc/version/boot)
+	vecho "    boot image " "$(cat etc/version/boot)"
 
 	rm -f platform
-	ln -s ./platform-$pistamp platform
+	ln -s "./platform-$pistamp" platform
 }
 
 remove() {
 	pistamp=$1
-	piname_present_get_bootfs $pistamp $2
-	cd /$bootfs
+	piname_present_get_bootfs "$pistamp" "$2"
+	cd "/$bootfs" || fatal "Could not chdir to /$bootfs"
 	bootstamp=$(cat platform/etc/version/platform)
 
 	if [[ -d platform-$pistamp ]]; then
-		if [[ $bootstamp == $pistamp ]]; then
+		if [[ $bootstamp == "$pistamp" ]]; then
 			eecho "$pistamp is the next-booting PI." \
 		    		"Please activate another PI"
 			eecho "using 'piadm activate <other-PI-stamp>' first."
 			usage
 		fi
 
 		# Boot image processing.
-		if [[ -d boot-$pistamp ]]; then
+		if [[ -d "boot-$pistamp" ]]; then
 			# Boot bits may be older than the current PI, and the
 			# current PI may not have matching boot bits for some
 			# reason. Guard against shooting yourself in the foot.
-			grep -q $pistamp etc/version/boot
-			if [[ $? -eq 0 ]]; then
+			if ! grep -q "$pistamp" etc/version/boot; then
```

Comment:
This is a place where you should lose the '!'.  If the grep matches, we utter the warning and bail.

Reply:
This should be good